### PR TITLE
Add model changes for real classes

### PR DESCRIPTION
**Pull request by Blu3spirits**
This is the 'staging' commit for all the work that needs to be finished
before we can implement the CourseShell model.

Notes:
- Renamed Course model to TransferCourse
- Changed access url to `/api/transfer-courses`
- The new url endpoint will be `/api/courses` for the joined viewset
- Added base model for real MTU classes
- This WILL break deployments since the sync scripts cannot yet write to
   the correct model. The scrape script for MTU classes needs to be
   finished first.

### CODE REVIEW

File: course_gather/models.py [hunk 1]
```diff
@@ -48,7 +48,45 @@ class Meta:
...
 
-class Course(models.Model):
+class ShellCourse(models.Model):
+    '''
+    Shell object of courses that is the intermediary between
+    Courses and MTUCourses
+    TransferCourse <=> Shell <=> MTUCourse
+
+    In order for this to work, a shell is going to have to related to many
+    Courses and only One MTUCourse
```
Comment (Blu3spirits) [inline course_gather/models.py]: This is a double MtM relationship actually. The comment is outdated

File: course_gather/models.py [hunk 2]
```diff
@@ -48,7 +48,45 @@ class Meta:
...
+    Example:
+    Cyber Sec (Two entries)
+
+    There's only one cyber sec at MTU
+    MTM<================>MTM
+    '''
```
Comment (Blu3spirits) [inline course_gather/models.py:56]: There's some additional notes about this, the main point though is that there will be data loss when we join these two data sets. Namely the strange $COURSE_CODE 1XXX courses that are pulled in from places. Along with "Unassigned transfers"

File: course_gather/models.py [hunk 3]
```diff
@@ -48,7 +48,45 @@ class Meta:
...
+        return self.shell_name
+
+    class Meta:
+        ordering = ['shell_name']
+        verbose_name = 'course_shell'
+        verbose_name_plural = 'course_shells'
+        db_table = 'course_shells'
+
+
+class TransferCourse(models.Model):
```
Comment (Blu3spirits) [inline course_gather/models.py:98]: None of these model fields are in any way finalized so if anyone has comments about what they could/should be lmk.

File: course_gather/models.py [hunk 4]
```diff
@@ -88,18 +126,55 @@ class MTUCourse(models.Model):
...
+    mtu_course_location = models.CharField(
+                            max_length=10,
+                            help_text='The MTU building location code the \
+                            class is located in')
+    mtu_course_fee = models.CharField(
+                            max_length=80,
+                            help_text='The fee associated with the course')
 
     def __str__(self):
         return self.mtu_course_id
```
Comment (codetheweb) [inline course_gather/models.py:194]: Are we still planning to use a composite primary key?

File: course_gather/models.py [hunk 5]
```diff
@@ -88,18 +126,55 @@ class MTUCourse(models.Model):
...
+                            max_length=40,
+                            help_text='The hours the MTU course runs')
+    mtu_course_capacity = models.IntegerField(
+                            help_text='The capacity of the class')
+    mtu_course_capacity_enrolled = models.IntegerField(
+                            help_text='The amount of students currently \
+                            enrolled in the course')
+    mtu_course_capacity_remaining = models.IntegerField(
+                            help_text='The seats remaining in the course')
+    mtu_course_instructor = models.CharField(
```
Comment (codetheweb) [inline course_gather/models.py:173]: Are we going to have a separate table for instructors, or is that out of the scope of this PR?

File: course_gather/models.py [hunk 6]
```diff
@@ -88,18 +126,55 @@ class MTUCourse(models.Model):
...
+                            teaching the course')
+    mtu_course_duration = models.CharField(
+                            max_length=30,
+                            help_text='The date range the class runs')
+    mtu_course_location = models.CharField(
+                            max_length=10,
+                            help_text='The MTU building location code the \
+                            class is located in')
+    mtu_course_fee = models.CharField(
+                            max_length=80,
```
Comment (codetheweb) [inline course_gather/models.py]: I'm pretty sure this value is always in dollars, so could we store this in cents instead of a string?

**Review by codetheweb — CHANGES_REQUESTED** [review]
I'm having some trouble understanding the Shell model. Why is it a many-to-many relation between Shell and MTUCourse?

File: course_gather/models.py (hunk 4 above)
Comment (Blu3spirits) [reply course_gather/models.py:194]: Yep, we're just going to need that data from the scraper first

> I'm having some trouble understanding the Shell model. Why is it a many-to-many relation between Shell and MTUCourse?

TL;DR, there can be 0, 1, or many classes with a given name, and there can be 0, 1, or many transfer classes that can apply to the MTU classes.

The biggest reason for this is that there are often times multiple sections associated with an MTU course CRN. For example, Physics By Inquiry II has 17 sections. This Shell model is meant to represent the name of the course so we can relate those 17 MTU classes to the 89 transfer courses that are in the current dataset.

File: course_gather/models.py (hunk 5 above)
Comment (Blu3spirits) [reply course_gather/models.py:173]: Out of scope at the moment. There are some courses that are taught by upwards of 8 professors and have all of their names on it.

File: course_gather/models.py (hunk 6 above)
Comment (Blu3spirits) [reply course_gather/models.py]: It is almost always this field,
`Lab/Course Fee: $XXX.XXX `

We could eventually try our hand at parsing this value out, but that will have to happen in the scraper.

File: course_gather/models.py (hunk 5 above)
Comment (Blu3spirits) [reply course_gather/models.py:173]: Here's an example of such a course:
Instructor(s):
```

  | Jarvi , Cavaleri , Fiss , Flaspohler , Hyslop , Rudnicki , Schmierer
-- | --

```
And it is this class:
```

83435 | FW | 4800 | 0A | 1 | 2.000 | Commun. for Natural Resources
-- | -- | -- | -- | -- | -- | --



```

File: course_gather/models.py (hunk 4 above)
Comment (codetheweb) [reply course_gather/models.py:194]: Got it.

In that case, would it make sense to omit the `name` field from MTUCourse since it's duplicated in Shell instances?

File: course_gather/models.py (hunk 6 above)
Comment (codetheweb) [reply course_gather/models.py]: Right, but since the scrapper isn't included in this PR might as well make it cents now. 🤷 

I guess it doesn't really matter that much.

File: course_gather/models.py (hunk 4 above)
Comment (Blu3spirits) [reply course_gather/models.py:194]: Not really because we're probably going to want that data for the front end, also imo it makes sense to keep the duplicated data in the object. This object structure is more or less just for easy searching UI/API wise. But if we omit the name instance, while we're in our loops we lose some context unless we go out of the way to track the top level

Comment (Blu3spirits) [reply course_gather/models.py:194]: That said, later on once we actually get data and start work on the UI and get that working, if we want to remove it, it's a simple fix :man_shrugging: .

File: course_gather/models.py (hunk 6 above)
Comment (Blu3spirits) [reply course_gather/models.py]: I just had an idea. We could preserve this data and make a real price column and have a tags system akin to the chips idea you originally posted. There exists other entries in the fee column such as:
```
On Line Technology Fee $114.00 and Lab/Course Fee: $20.00 
Lab/Course Fee:/ENG4910 $50.00
On Line Technology Fee $114.00 
Lab/CourseFee/ENG4900 $162.00
Lab/Course Fee: P1000 $15.00
```

Comment (Blu3spirits) [reply course_gather/models.py]: Going to do that and see how well it goes

File: course_gather/models.py (hunk 4 above)
Comment (codetheweb) [reply course_gather/models.py:194]: 👍

**Review by codetheweb — APPROVED** [review]

File: course_gather/models.py (hunk 6 above)
Comment (Blu3spirits) [reply course_gather/models.py]: 0335662

**Comment by Blu3spirits** [comment]
> Usually USD values are stored in databases as cents to avoid rounding errors; but I guess it doesn't matter as much here since we're not calculating anything off them. Up to you.

I hadn't heard that before but I see the sense in it. However looking at the data MTU stores the data with only two decimals anyway so really I don't see any issues coming about from this. I do think that eventually with the cart system a "total fees" would be useful to have. If it does become an issue ever we can switch over. Or I can increase the floating-point limits.

**Comment by Blu3spirits** [comment]
> That being said, would using an enum or something for this field make more sense? I guess it doesn't have to be here but we should have a static list of possible tags somewhere so we can map them to values if necessary. I.e. tags with value 'Online' we may want to display with a yellow background.

I don't really think an enum would make sense since it would require a code and database change for a new tag. With this current tag model, it's very arbitrary on purpose so that we can actually do things like labels based on location. In the sync script that feeds from the scraper we just do a update_or_create() model method that creates a new label and applies to the current object. Since that bit of `if location == 'online'` would have to be handled it would effectively end up being static list anyway. 

I think farther down the line as well after we collect data from rate my professor we could also add a tag to show that.
